File: Lab6/Parser.java
import domain.Pair;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Parser {

    private final Grammar grammar;

    public Parser() {
        this.grammar = new Grammar();
    }

    public List<String> getN() {
        return grammar.getN();
    }

    public List<String> getE() {
        return grammar.getE();
    }

    public List<Pair<String, String>> getP() {
        return grammar.getP();
    }

    public void readGrammar(String filename) throws Exception {
        grammar.readGrammar(filename);
    }

    public List<Pair<String, String>> ClosureLR(String analysis) {
        List<String> tokens = Arrays.stream(analysis.strip()
                .replace("->", " ")
                .split(" "))
                .collect(Collectors.toList());
        List<Pair<String, String>> P = new ArrayList<>();
        P.add(new Pair<>(tokens.get(0), tokens.get(1)));
        int index;
        char nonT;
        int size = 1;

        do {
            size = P.size();
            List<Pair<String, String>> filteredP = new ArrayList<>(P);
            for (Pair<String, String> production : filteredP) {
                index = production.getValue().indexOf('.');
                if (index != -1 && index < production.getValue().length() - 1) {
                    nonT = production.getValue().charAt(index + 1);
                    List<Pair<String, String>> filteredB = grammar.filterP(Character.toString(nonT));
                    for (Pair<String, String> productionB : filteredB) {
                        if (!P.contains(new Pair<>(productionB.getKey(), "." + productionB.getValue()))) {
                            P.add(new Pair<>(productionB.getKey(), "." + productionB.getValue()));
                        }
                    }
                }
            }

        } while (size < P.size());
        return P;
    }


    public List<Pair<String, String>> gotoLR(List<Pair<String, String>> productions, String symbol) {
        List<Pair<String, String>> nestedList = new ArrayList<>();
        productions.stream()
                .filter(item -> item.getValue().contains("." + symbol))
                .map(item -> new Pair<>(item.getKey(), item.getValue().replace("." + symbol, symbol + ".")))
                .forEach(item -> nestedList.addAll(ClosureLR(item.getKey() + "->" + item.getValue())));
        return nestedList.stream().distinct().collect(Collectors.toList());


    }

    public List<List<Pair<String, String>>> ColCan_LR() {

        List<List<Pair<String, String>>> C = new ArrayList<>();

        C.add(ClosureLR("S'->.S"));

        boolean dirty;

        do {
            dirty = false;
            List<List<Pair<String, String>>> filteredC = new ArrayList<>(C);
            for (List<Pair<String, String>> state : filteredC) {
                for (String element : Stream.concat(getN().stream(), getE().stream())
                        .collect(Collectors.toList())) {

                    List<Pair<String, String>> goToRes = gotoLR(state, element);

                    if (!goToRes.isEmpty() && !includedForEach(C, goToRes)) {
                        C.add(goToRes);
                        dirty = true;
                    }
                }
            }
        } while (dirty);

        return C;
    }

    private boolean includedForEach(List<List<Pair<String, String>>> C, List<Pair<String, String>> goToRes) {
        return C.stream().anyMatch((listOfStates) -> included(listOfStates, goToRes));
    }

    private boolean included(List<Pair<String, String>> C, List<Pair<String, String>> goToRes) {
        return C.containsAll(goToRes);
    }

    public HashMap<Integer, Pair<String, HashMap<String, Integer>>> createLRTable(List<List<Pair<String, String>>> states) {
        HashMap<Integer, Pair<String, HashMap<String, Integer>>> lrTable = new HashMap<>();

        for (List<Pair<String, String>> state : states) {
            int position=states.indexOf(state);
            if(state.contains(new Pair("S'","S."))){
                lrTable.put(position,new Pair("acc",new HashMap<String,Integer>()));
            }
            else if(hasReduce(state)!=-1){
                lrTable.put(position,new Pair("reduce "+hasReduce(state),new HashMap<String,Integer>()));
            }
            else {
                lrTable.put(position,new Pair("shift",new HashMap<String,Integer>()));
                for (String element : Stream.concat(getN().stream(), getE().stream())
                        .collect(Collectors.toList())) {

                    List<Pair<String, String>> goToRes = gotoLR(state, element);

                    if (!goToRes.isEmpty()) {
                        lrTable.get(position).getValue().put(element,states.indexOf(goToRes));

                    }
                }
            }
        }


        return lrTable;


    }



    private Integer hasReduce(List<Pair<String, String>> state) {
        for(Pair<String,String> lrItem: state){
            if(lrItem.getValue().charAt(lrItem.getValue().length()-1)=='.'){
                return grammar.getP().indexOf(new Pair(lrItem.getKey(),lrItem.getValue().substring(0,lrItem.getValue().length()-1)));
            }
        }
        return -1;
    }

    public List<String> parsingAlg(HashMap<Integer, Pair<String, HashMap<String, Integer>>>  lrTable,List<List<Pair<String, String>>> C,  String word) {

        //TODO work with stack
        List<Pair<String, String>> state = C.get(0);
        List<String> alpha = new ArrayList<>();
        ;
        List<Character> beta = new ArrayList<>();
        List<String> phi = new ArrayList<>();
        alpha.add("0");
        for (int i = 0; i <= word.length()-1; i++) {
            beta.add(word.charAt(i));
        }
        while(true){
            int position=Integer.parseInt(alpha.get(alpha.size()-1));
            if (lrTable.get(position).getKey().equals("shift")) {
                Character a = beta.remove(0);
                state = gotoLR(state, a.toString());
                alpha.add(a.toString());
                alpha.add(Integer.toString(C.indexOf(state)));
            } else if (lrTable.get(position).getKey().contains("reduce")) {
                String findReducer=lrTable.get(position).getKey().substring(7);
                int reducer=Integer.parseInt(findReducer);
                Pair<String,String> production=search_prod(reducer);
                for(int i=0;i<2*production.getValue().length();i++) {
                    alpha.remove(alpha.size()-1); //?
                }
                state = gotoLR(C.get(Integer.parseInt(alpha.get(alpha.size()-1))), production.getKey());
                alpha.add(Character.toString(production.getKey().charAt(0)));
                alpha.add(Integer.toString(C.indexOf(state)));
                phi.add(findReducer);
            }
            else{
                if(lrTable.get(position).getKey().equals("acc")){
                    Collections.reverse(phi);
                    return phi;
                }
                else{
                    return null;
                }

            }
        }
    }

    private Pair<String, String> search_prod(int reducer) {
        return grammar.getP().get(reducer);
    }

    public void createTable(List<String> productions){
        List<Pair<String,Pair<Integer,Integer>>> table=new ArrayList<>();
        table.add(new Pair<>("S",new Pair<>(-1,-1)));
        int k=0;
        for(String production:productions){
            while(!grammar.getN().contains(table.get(k).getKey())){
                k++;
            }
            Pair<String, String> prod=search_prod(Integer.parseInt(production));
            int index=0;
            table.add(new Pair(String.valueOf(prod.getValue().charAt(0)),new Pair(k,-1)));
            index++;
            for(;index<prod.getValue().length();index++){
                table.add(new Pair(String.valueOf(prod.getValue().charAt(index)),new Pair(k,k+index)));
            }
            k++;
        }
        System.out.println(table);
    }
}

File: Lab6 redone/Parser.py
from typing import Optional

from Grammar import *

class Parser:
    def __init__(self):
        self._grammar = Grammar()

    @property
    def N(self) -> list:
        return self._grammar.N

    @property
    def E(self) -> list:
        return self._grammar.E

    @property
    def P(self) -> list:
        return self._grammar.P

    def read_grammar(self, filename: str):
        self._grammar.read_grammar(filename)

    @staticmethod
    def find_string(lst: list, string: str) -> int:
        for i in range(len(lst)):
            if lst[i] == string:
                return i
        return -1

    def closure_LR(self, analysis: Pair) -> list:
        tokens = [analysis.first, analysis.second]
        P = list()
        P.append(Pair(tokens[0], tokens[1]))
        size = 0
        while size < len(P):
            size = len(P)
            filteredP = P.copy()
            for pair in filteredP:
                index = self.find_string(pair.second.split(" "), '.')
                if index != -1 and index < len(pair.second.split(" ")) - 1:
                    nonT = pair.second.split(" ")[index + 1]
                    filteredB = self._grammar.filter_P(nonT)
                    filteredB = map(lambda p: Pair(p.first, '. ' + p.second), filteredB)
                    for pairB in filteredB:
                        if pairB not in P:
                            P.append(pairB)
        return P

    def goto_LR(self, productions: list, symbol: str) -> list:
        nestedList = []
        productions = map(lambda el: Pair(el.first, el.second.replace(". " + symbol, symbol + " .")),
                          filter(lambda el: (". " + symbol) in el.second, productions))
        for pair in productions:
            nestedList.extend(self.closure_LR(pair))
        return nestedList

    def includedForEach(self, C: list, goToRes: list) -> bool:
        return any(self.included(states, goToRes) for states in C)

    @staticmethod
    def included(C: list, goToRes: list) -> bool:
        return all(elem in C for elem in goToRes)

    def colcan_LR(self) -> list:
        C = list()
        C.append(self.closure_LR(Pair("S'", ". S")))
        dirty = True
        while dirty:
            dirty = False
            filteredC = C.copy()
            for state in filteredC:
                for elem in self.N + self.E:
                    goToRes = self.goto_LR(state, elem)
                    print(self.goto_LR(state, elem))

                    if len(goToRes) != 0 and not self.includedForEach(C, goToRes):
                        C.append(goToRes)
                        dirty = True
        return C

    def _has_reduce(self, state: list) -> int:
        for el in state:
            if el.second.split(" ")[-1] == '.':
                return self._grammar.P.index(Pair(el.first, el.second[0:-2]))
        return -1

    def create_LR_table(self, states: list) -> dict:
        lrTable = {}

        for position, state in zip(range(len(states)), states):
            if Pair("S'", "S .") in state:
                lrTable[position] = Pair("acc", {})
            elif (self._has_reduce(state)) != -1:
                lrTable[position] = Pair("reduce " + str(self._has_reduce(state)), {})
            else:
                lrTable[position] = Pair("shift", {})
                for el in self.N + self.E:
                    goToRes = self.goto_LR(state, el)
                    if len(goToRes) != 0:
                        lrTable[position].second[el] = states.index(goToRes)


        return lrTable

    def parsing_alg(self, lrTable: dict, C: list, word: str) -> Optional[list]:
        state = C[0]
        alpha = []
        beta = []
        phi = []
        alpha.append("0")
        parse = word.split(" ")
        for p in parse:
            beta.append(p)

        while True:
            position = int(alpha[-1])
            if lrTable[position].first == "shift":
                a = beta.pop(0)
                state = self.goto_LR(state, a)
                alpha.append(a)
                alpha.append(str(C.index(state)))
            elif lrTable[position].first == "reduce":
                phi.append(lrTable[position].first[:7])
                reducer = int(phi[-1])
                production = self.P[reducer]
                for i in range(2 * len(production.second.split(" "))):
                    alpha.pop()
                state = self.goto_LR(C[int(alpha[-1])], production.first)
                alpha.append(production.first)
                alpha.append(str(C.index(state)))
            else:
                if lrTable[position].first == "acc":
                    return list(reversed(phi))
                return None


    def create_table(self, productions: list) -> list:
        table = list()
        table.append(Pair("S", Pair(-1, -1)))
        kStack = []
        k = 0
        index = 0
        kStack.append(k)
        for production in productions:
            k = kStack.pop()
            while table[k].first not in self.N:
                k += 1
                index += 1
            prod = self.P[int(production)]
            i = 0
            if prod.second.split(" ")[i] in self.N:
                kStack.append(index + 1)
            index += 1
            table.append(Pair(prod.second.split(" ")[i], Pair(k, -1)))
            i += 1
            while i < len(prod.second.split(" ")):
                if prod.second.split(" ")[i] in self.N:
                    kStack.append(index + 1)
                index += 1
                table.append(Pair(prod.second.split(" ")[i], Pair(k, index)))
        return table
